### PR TITLE
Scattering-diffuse-mni

### PR DESCRIPTION
**Pull request by MaxNilo**
### Which issue(s) are closed by this pull request?

Closes #

### Changes proposed in this pull request:

-
-
-

### Labels

Label your issue to make it easier for us to assign and track:

Use one of these labels:
- **hot:** For bugs on the master branch
- **bug:** For bugs not on the master branch
- **enhancement:** For suggesting enhancements of current functionality
- **feature:** For requesting new features
- **documentation:** Everything related to docstrings and comments

### CODE REVIEW

File: tests/test_diffuse.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+import imkar as ik
+import numpy as np
+import pyfar as pf
+import pytest
+import numpy.testing as npt
+
+
```
Comment (ahms5) [inline tests/test_diffuse.py:7]: add test for multidimensional inputs
np.vstack(([T60_1, T60_2, T60_3, T60_4], [T60_1, T60_2, T60_3, T60_4]))

File: imkar/scattering/diffuse.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        raise ValueError("reverb_time.cshape has to be (4,)")
+
+    if len(speed_of_sound) != 4:
+        raise ValueError("speed_of_sound has to be a vector containing the \
+                         speed of sound during all 4 measurement conditions")
+
+    if (np.any(np.asarray(speed_of_sound, dtype=float) <= 0)):
+        raise ValueError("speed_of_sound has to be a value bigger than zero")
+
+    if type(air_attenuation) == pf.classes.audio.FrequencyData:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:79]: if isinstance(air_attenuation, pf.FrequencyData):

File: imkar/scattering/diffuse.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        air_attenuation = np.asarray(air_attenuation, dtype=float)
+    if air_attenuation.shape != (4, 1, reverb_time.n_bins):
+        raise ValueError("air_attenuation has to be of shape (4, 1,\
+                         number_of_frequencies)")
+
+    if (np.any(air_attenuation < 0)):
+        raise ValueError("air_attenuation has to be a value bigger or equal \
+                         to zero")
+
+    if (calculation_method != "ISO" and calculation_method != "Eyring" and
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:91]:
```suggestion
    if calculation_method  in ['ISO', 'Eyring', 'Sabine']:
```

**Review by MaxNilo — COMMENTED** [review]
Suggestions applied and addition of multidimensional input.

File: imkar/scattering/diffuse.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            surface_room=9.05, volume_room=1.67):
+    """
+    This function calculates the diffuse scattering coefficient after the ISO
+    17497-1 method. It uses four reverberation times measured under different
+    conditions [#]_.
+
+    Parameters
+    ----------
+    reverberation_time : pf.FrequencyData
+        Frequency object containing the reverbaration times calculated during
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:18]:
```suggestion
        Frequency object containing the reverberation times calculated during
```

File: imkar/scattering/diffuse.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        measurement condition calculated by temperature and air humidity during
+        measurement time. Its cshape has to be (..., 4).
+
+    calculation_method : 'ISO','Eyring' or 'Sabine'
+        Defines the method by which alpha is calculated.
+        For method 'ISO' the baseplate_area and sample_area need to be equal.
+
+    scale_factor : float
+        Factor that corrects the frequencies when using a smaller scale
+        reverbaration room than defined in the ISO standard. Standard value is
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:39]:
```suggestion
        reverberation room than defined in the ISO standard. Standard value is
```

File: imkar/scattering/diffuse.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        Array containing the speed of sound in m/s for every measurement
+        condition calculated by the temperature and air humidity during
+        measurement time.
+
+    air_attenuation : array, pf.FrequencyData
+        Matrix or FrequencyData containing the air attenuation in 1/m for every
+        measurement condition calculated by temperature and air humidity during
+        measurement time. Its cshape has to be (..., 4).
+
+    calculation_method : 'ISO','Eyring' or 'Sabine'
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:33]:
```suggestion
    calculation_method : string
```
describe each option as it is done for distance_measure in find_nearest
https://pyfar.readthedocs.io/en/stable/classes/pyfar.coordinates.html#pyfar.classes.coordinates.Coordinates.find_nearest

File: imkar/scattering/diffuse.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+    if calculation_method not in ['ISO', 'Eyring', 'Sabine']:
+        raise ValueError("Use 'ISO', 'Eyring' or 'Sabine' as calculation \
+                         method")
+
+    if (calculation_method == "ISO") & (sample_area != baseplate_area):
+        raise ValueError("For calculation method 'ISO' the baseplate_area and \
+                         sample_area need to be equal.")
+
+    if not isinstance(scale_factor, numbers.Real) or scale_factor <= 0:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:112]:
```suggestion
    if not isinstance(scale_factor, (float, int)) or scale_factor <= 0:
```
in this way we dont need numbers package anymore

File: imkar/scattering/diffuse.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        raise TypeError("scale_factor has to be a real number >0")
+
+    if not isinstance(sample_area, numbers.Real) or sample_area <= 0:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:115]: see above

File: imkar/scattering/diffuse.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        raise TypeError("sample_area has to be a real number >0")
+
+    if not isinstance(baseplate_area, numbers.Real) or baseplate_area <= 0:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:118]: see above

File: imkar/scattering/diffuse.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        raise TypeError("baseplate_area has to be a real number >0")
+
+    if not isinstance(surface_room, numbers.Real) or surface_room <= 0:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:121]: see above

File: imkar/scattering/diffuse.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        raise TypeError("surface_room has to be a real number >0")
+
+    if not isinstance(volume_room, numbers.Real) or volume_room <= 0:
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:124]: see above

File: imkar/scattering/diffuse.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+        the four different measurements. If the measurement is performed at
+        multiple positions, the mean value for every condition of the
+        measurements is taken. Its cshape has to be (..., 4).
+
+    speed_of_sound : array, float
+        Array containing the speed of sound in m/s for every measurement
+        condition calculated by the temperature and air humidity during
+        measurement time.
+
+    air_attenuation : array, pf.FrequencyData
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:28]: don't allow array, it's frequency data, so force pf.FrequencyData
```suggestion
    air_attenuation : pf.FrequencyData
```

File: imkar/scattering/diffuse.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        raise ValueError("speed_of_sound has to be an array containing the \
+                         speed of sound during all 4 measurement conditions. \
+                         Its shape has to be (..., 4) and equal to \
+                         reverberation_time.cshape.")
+
+    if (np.any(np.asarray(speed_of_sound, dtype=float) <= 0)):
+        raise ValueError("speed_of_sound has to be a value bigger than zero")
+
+    if isinstance(air_attenuation, pf.FrequencyData):
+        air_attenuation = np.asarray(air_attenuation.freq, dtype=float)
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:91]: dont convert to array, use pf.FrequencyData

File: imkar/scattering/diffuse.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+    if (speed_of_sound.shape != reverberation_time.cshape):
+        raise ValueError("speed_of_sound has to be an array containing the \
+                         speed of sound during all 4 measurement conditions. \
+                         Its shape has to be (..., 4) and equal to \
+                         reverberation_time.cshape.")
+
+    if (np.any(np.asarray(speed_of_sound, dtype=float) <= 0)):
+        raise ValueError("speed_of_sound has to be a value bigger than zero")
+
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:89]: add a check if frequencies of ``air_attenuation``and ``reverberation_times``are same.

File: imkar/scattering/diffuse.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        # specular absorption coefficient
+        a_spec = surface_room/sample_area*(alpha[:, 3]-alpha[:, 2])+alpha[:, 2]
+
+        # scattering coefficient for the base plate
+        s_base = surface_room/baseplate_area * \
+            (alpha[:, 2]-alpha[:, 0]).clip(min=0)
+
+    # random-incidence scattering coefficient
+    scatterring_coefficient = (1-(1-a_spec)/(1-a_s)).clip(min=0)
```
Comment (ahms5) [inline imkar/scattering/diffuse.py:178]:
```suggestion
    scattering_coefficient = (1-(1-a_spec)/(1-a_s)).clip(min=0)
```

**Review by ahms5 — COMMENTED** [review]
Thank you for appling the changes. I merged the latest version into the branch and added the docs. I have a few further small comments. If you still have time, please include it.

**Comment by ahms5** [comment]
replaced by #23

**Pull request closed without merging**